src/modules/mc_hover_thrust_estimator/MulticopterHoverThrustEstimator.cpp

```diff
@@ -105,42 +110,55 @@ void MulticopterHoverThrustEstimator::Run()
 
 	perf_begin(_cycle_perf);
 
-	vehicle_land_detected_s vehicle_land_detected;
-	vehicle_status_s vehicle_status;
-	vehicle_local_position_s local_pos;
+	if (_vehicle_land_detected_sub.updated()) {
+		vehicle_land_detected_s vehicle_land_detected;
 
-	if (_vehicle_land_detected_sub.update(&vehicle_land_detected)) {
-		_landed = vehicle_land_detected.landed;
+		if (_vehicle_land_detected_sub.copy(&vehicle_land_detected)) {
+			_landed = vehicle_land_detected.landed;
+		}
 	}
 
-	if (_vehicle_status_sub.update(&vehicle_status)) {
-		_armed = (vehicle_status.arming_state == vehicle_status_s::ARMING_STATE_ARMED);
+	if (_vehicle_status_sub.updated()) {
+		vehicle_status_s vehicle_status;
+
+		if (_vehicle_status_sub.copy(&vehicle_status)) {
+			_armed = (vehicle_status.arming_state == vehicle_status_s::ARMING_STATE_ARMED);
+		}
 	}
 
-	if (_vehicle_local_pos_sub.update(&local_pos)) {
+	// always copy latest position estimate
+	vehicle_local_position_s local_pos{};
+
+	if (_vehicle_local_pos_sub.copy(&local_pos)) {
 		// This is only necessary because the landed
 		// flag of the land detector does not guarantee that
 		// the vehicle does not touch the ground anymore
 		// TODO: improve the landed flag
-		_in_air = local_pos.dist_bottom > 1.f;
+		if (local_pos.dist_bottom_valid) {
+			_in_air = local_pos.dist_bottom > 1.f;
+		}
 	}
 
-	ZeroOrderHoverThrustEkf::status status{};
-
-	if (_armed && !_landed && _in_air) {
+	if (_armed && !_landed && _in_air && (local_pos.timestamp > _timestamp_last)) {
 		vehicle_local_position_setpoint_s local_pos_sp;
 
 		if (_vehicle_local_position_setpoint_sub.update(&local_pos_sp)) {
 
-			const hrt_abstime now = hrt_absolute_time();
-			const float dt = math::constrain((now - _timestamp_last) / 1e6f, 0.002f, 0.2f);
+			const float dt = (local_pos.timestamp - _timestamp_last) * 1e-6f;
+			_timestamp_last = local_pos.timestamp;
+
+			if ((dt > 0.001f) && (dt < 0.1f)) {
```

Comment:
Is a protection against to low dt needed?

Reply:
If dt is too small or too big, we do not propagate the variance, but update the `timestamp_last`. So we skip to increase the variance, is this indented?

Reply:
I don't think it's needed, but I also didn't really have justification for purging it. Things would have to be really badly broken for it to matter.

Reply:
I think the only real case that might matter is switching out of a position control mode and then back in later. The `dt` on the first iteration back would then be the time spent in a non position control mode. Should we explicitly handle it here?

Reply:
I'll keep the upper limit, because that could actually catch transitions from not running to activation.